clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -302,27 +302,20 @@ private long sendProducerData(long now) {
     }
 
     private boolean maybeSendTransactionalRequest(long now) {
-        if (transactionManager.isCompletingTransaction() &&
-                !transactionManager.hasPartitionsToAdd() &&
```

Comment:
@hachikuji is correct. when we were completing a transaction with unflushed batches, we needed to ensure that the corresponding `AddPartitions` request would go out before the batches themselves. Hence this check.

With the current patch, this is transparently handled by the `nextRequestHandler` call, which will ensure the send order is `AddPartitions`, `flush batches`, `EndTxn`.